lua/telescope/builtin/git.lua

```diff
@@ -10,16 +10,25 @@ local conf = require('telescope.config').values
 local git = {}
 
 git.files = function(opts)
-  local show_untracked = utils.get_default(opts.show_untracked, true)
+  local show_untracked = utils.get_default(opts.show_untracked, false)
+  local recurse_submodules = utils.get_default(opts.recurse_submodules, false)
 
   -- By creating the entry maker after the cwd options,
   -- we ensure the maker uses the cwd options when being created.
   opts.entry_maker = opts.entry_maker or make_entry.gen_from_file(opts)
 
+  local args = { "git", "ls-files", "--exclude-standard", "--cached" }
+  if show_untracked then
+    table.insert(args, "--others")
+  end
+  if recurse_submodules then
+    table.insert(args, "--recurse-submodules")
+  end
+
   pickers.new(opts, {
     prompt_title = 'Git File',
     finder = finders.new_oneshot_job(
-      { "git", "ls-files", "--exclude-standard", "--cached", show_untracked and "--others" },
+      args,
```

Comment:
Lets just do this in place, similar to what we had before. `recurse_subodules and "--recurse-submodules"`. 
You can split up the table over multiple lines, if CI doesn't like the length

Reply:
I started with that, but I think if the `show_untracked` condition is false,  the `recurse_submodule` one is not passed to `git`. Possibly because the function stops getting arguments when finding a `false`. Can that be?

Reply:
You are right. Ignore that :)

Reply:
Okay it actually does work. You just have to do `show_untracked and "--others" or nil`. If you don't do this that would put `false` into the table. After that the table looks like that:
```
{ "git", "ls-files", "--exclude-standard", "--cached",
  [6] = "--recurse-submodules"
}
```
Now all we have to do is `vim.tbl_flatten`. Put that in front of the table. After that `[5]` will no longer be `nil`. Now its `--recurse-submodules`.

Reply:
I see! Question: would it be better to use `tbl_flatten` directly inside the `oneshot` function (I forget the name now), or maybe even discard `false` values to avoid using it at all? Semantically it would seam to make sense, and I'm assuming this kind of pattern might recur often

Reply:
We have
```lua
    finder = finders.new_oneshot_job(
      vim.tbl_flatten { conf.vimgrep_arguments, opts.word_match, search},
      opts
    ),
```
for `live_grep` for example. So just define your table as argument for `vim.tbl_flatten`.

Reply:
Sorry it's taking me a while, I was on another PC.

Another small roadblock. `utils.get_default` seems problematic. If I run
`:Telescope git_files show_untracked=false` or `:Telescope git_files show_untracked=0`, that parameter is treated as a string, not a bool, and so when I do `show_untracked and "--others" or nil` I still get `"--others"`.

Should the ex-command invocation syntax be different?

Reply:
No problem. Take all the time you need.
Oh that is an interesting problem. I never use `:Telescope`, only `:lua require` but we should fix that, probably inside utils.get_default. You wanna do it? And thanks for discovering this one :)

Reply:
Hmmm I wonder how to approach this.

We might still want to pass `false` as a string in some cases, so automatically converting the string to the bool feels wrong.

My hunch would be, for boolean flags, to accept these syntaxes:
```vim
:Telescope git_files show_untracked
:Telescope git_files !show_untracked
```
and convert them to `true` and `false` respectively.

But I get an error:
```
Error detected while processing function <SNR>54_load_command:
line   11:                                                                                                                                        
E684: list index out of range: 1
E15: Invalid expression: opt[1]
```
which feels more "internal" than `telescope` :p What do you think? Is it even `vim`-ish?

Reply:
Or maybe, more simply, a utility function that converts the value;  and then the called function, which is the only one who knows what type to expect, calls the appropriate one (`utils.to_bool` or `utils.to_num`)?

Reply:
`:Telescope git_files !show_untracked` that would just translate to writing `:Telescope git_files false` its not associated with a key. Lets just ping @tjdevries here. I don't know how to fix that from vim side. My take would be have `utils.get_default` have to handle this one but like 90% of the builtin dont even use `utils.get_default` so those are broken as well.
I personally always just do: `opts.show_untracked = opts.show_untracked or true`.

Reply:
Yeah lets just do what you said. Did read that before writing my last comment. Make sure to check the type with `type()` so it does not run for bools

Reply:
So I'd say we can't do:
```
  local show_untracked = utils.get_default(utils.as_bool(opts.show_untracked), true)
```
because if no option is passed we get a `nil` in `as_bool` and we can only return `false`, which would prevent the default of `true` to count.
So we could switch to 
```
  local show_untracked = utils.as_bool(utils.get_default(opts.show_untracked, true))
```
which works but is a bit silly (the default is already a bool, so why convert it).

So I'd rather suggest that we check the default value, and if it is a bool, then convert the passed parameter to bool, too.

This seems to work for all 4 combinations of `true`/`false` except when both flags are `true`, in which case I get:
```
[bat error]: '/home/t.../fatal: ls-files --recurse-submodules unsupported mode': No such file or directory (os error 2)
```

No idea where that comes from. Is `opts` being passed around too much? Should I remove the `opt` that are only for `git` before passing it elsewhere?

Reply:
I would say `utils.as_bool` could return a nil. When passed in a value that does not convert or passed in nil.
I would say just push what you have right now.
I will look at it later. I can't right now. But i would say because its a [bat error] its the previewer.

Reply:
Okay that is a git error. Just try running: `git ls-files --exclude-standard --cached --others --recurse-submodules`.
`--others` and `--recurse-submodules` does not work together.

Reply:
Wow I thought I tried that. Weird though, I don't see the issue myself :P 

Anyway, I'll fix `as_bool` to pass `nil` through. 
Do you also want it to accept only `true`, `false`, `"true"`, `"false"`, `0`, `1`, `"0"`, `"1"` (or a subset of that) and return `nil` otherwise? I think standard `vim`'s behaviour would be "every string that does not start with a digit greater than 0 is falsy", but that won't work here because `"true"` need to be `true` (I think), so... I'll leave that choice to you :)

But that's another reason why this should probably be handled at a lower level

Reply:
I would say if type bool return input.
If type string then return true if "true", elif "false" then return false otherwise return nil.
if type int (i actually don't know that type 🤣 ) if 0 false, else true.
else return nil

Reply:
There :)

Reply:
Okay [man page](https://git-scm.com/docs/git-ls-files#Documentation/git-ls-files.txt---recurse-submodules) actually says that there is currently only support for `--cached` mode, so no `--others`. So we need to make sure only one of those is true after that i think i will just merge. Unless you are not yet done :)

Reply:
Hmm I thought about that, but
1. if they add support later, we'd have to remember and remove the check. Besides, it's the user's problem that they specifically ask for something that won't work, no? :P But you are right, they'd probably end up here opening an issue, just like I didn't know it either ^_^
2. not sure what to do. Just print and exit? Is there a pattern for echo-ing errors?

Reply:
We can't have it error out without a good error message.
So you could print a hint with `print(msg)` and just default to one of those or error out with `error(msg)`

Reply:
Done